doc/namespace.md

```diff
@@ -0,0 +1,417 @@
+
+#### Expose top level methods as a method of the namespace object
+
+Currently, top level methods in namespaces are not accessible from outside of the namespace. But there might be a use case to call other namespace's top level methods.
+
+#### Split root and builtin namespace
+
+NOTE: "builtin" namespace is a different one from the "builtin" namespace in the current implementation
+
+Currently, the single "root" namespace is the source of classext CoW. And also, the "root" namespace can load additional files after starting main script evaluation by calling methods which contain lines like `require "openssl"`.
+
+That means, user namespaces can have different sets of definitions according to when it is created.
+
+```
+[root]
+ |
+ |----[main]
+ |
+ |(require "openssl" called in root)
+ |
+ |----[ns1] having OpenSSL
+ |
+ |(remove_const called for OpenSSL in root)
+ |
+ |----[ns2] without OpenSSL
+```
+
+This could cause unexpected behavior differences between user namespaces. It should NOT be a problem because user scripts which refer to `OpenSSL` should call `require "openssl"` by themselves.
+But in the worst case, a script (without `require "openssl"`) runs well in `ns1`, but doesn't run in `ns2`. This situation looks like a "random failure" to users.
+
+An option possible to prevent this situation is to have "root" and "builtin" namespaces.
+
+* root
+  * The namespace for the Ruby process bootstrap, then the source of CoW
+  * After starting the main namespace, no code runs in this namespace
+* builtin
+  * The namespace copied from the root namespace at the same time with "main"
+  * Methods and procs defined in the "root" namespace run in this namespace
+  * Classes and modules required will be loaded in this namespace
+
+This design realizes a consistent source of namespace CoW.
+
+#### Separate cc_tbl and callable_m_tbl, cvc_tbl for less classext CoW
+
+The fields of `rb_classext_t` contains several cache(-like) data, `cc_tbl`(callcache table), `callable_m_tbl`(table of resolved complemented methods) and `cvc_tbl`(class variable cache table).
+
+The classext CoW is triggered when the contents of `rb_classext_t` are changed, including `cc_tbl`, `callable_m_tbl`, and `cvc_tbl`. But those three tables are changed by just calling methods or referring class variables. So, currently, classext CoW is triggered much more times than the original expectation.
+
+If we can move those three tables outside of `rb_classext_t`, the number of copied `rb_classext_t` will be much less than the current implementation.
+
+#### Object Shapes per namespace
+
+Now the classext CoW requires RClass and `rb_classext_t` to fallback its instance variable management from Object Shapes to the traditional `st_table`. It may have a performance penalty.
+
+If we can apply Object Shapes on `rb_classext_t` instead of `RClass`, per-namespace classext can have its own shapes, and it may be able to avoid the performance penalty.
```

Comment:
Though that probably means an overhead for the non-Namespace case as the Shape is at least one more indirection away from the Class object.